### ISSUE TITLE
xmlquery issue when values are set to match another variable

### ISSUE THREAD

**Issue by bartgol** [ESMCI/cime#4646]
I'm getting this error with --listall (or when query the HIST_N variable):

```bash
$ ./xmlquery HIST_N
ERROR: Entry HIST_N was listed as type int but value '$STOP_N' is not valid int
$ ./xmlquery --listall
ERROR: Entry HIST_N was listed as type int but value '$STOP_N' is not valid int
```

Did something change in CIME related to xmlquery? I don't think we recently changed the default for HIST_N, so I'm guessing this is a new thing? I'm using the tag cime6.0.226 (438850986).

**Comment by bartgol** [comment]
Ok, then maybe it's simply time to update the cime submodule in our fork. Thanks for checking!

**Comment by jgfouca** [comment]
It may be E3SM-only. I've assigned @jasonb5

**Comment by bartgol** [comment]
I may have a suspect. In generic_xml.py, we have

```python

        for m in reference_re.finditer(item_data):
            var = m.groups()[0]
            print (f"found m={m}, var={var}")
            logger.debug("find: {}".format(var))
            # The overridden versions of this method do not simply return None
            # so the pylint should not be flagging this
            # pylint: disable=assignment-from-none
            ref = self.get_value(var, subgroup=subgroup)
            ...
```
However, the referenced var may be in a different subgroup. Removing the subgroup arg from the `get_value` call, makes things work. However, that means that the referenced var must be unique across the XML file, which may be a restriction.

**Comment by rljacob** [comment]
I just tried ./xmlquery HIST_N on latest e3sm master with a case built back on May 14 and it works for me.

**Comment by jgfouca** [comment]
@rljacob , try this case: SMS.ne256pg2_ne256pg2.F2010-SCREAMv1

**Comment by rljacob** [comment]
Still works.   See /home/jacob/E3SM/cime/scripts/Fscream on Chrysalis.

**Comment by jgfouca** [comment]
Now I'm really confused.

**Comment by rljacob** [comment]
Is it an old case you're trying to query?   If you're doing that with a newer CIME, it may not work.

**Comment by jgfouca** [comment]
I just tried again, making sure i was on lastest master with subs updated, same error.

**Comment by rljacob** [comment]
Oh weird.  I tried actually doing "create_test" and going to the casedir and now I see the error.
Try making your case with: ./create_newcase --case Fscream --compset F2010-SCREAMv1 --res ne256pg2_ne256pg2 and going to the casedir it creates.

**Comment by jedwards4b** [comment]
It will only occur with create_test and not with create_newcase since the setting of
HIST_N=$STOP_N happens in config_tests.xml

**Comment by rljacob** [comment]
I went to a test directory created in our overnight testing and it was fine.  If it wasn't, all the tests would fail since there would be no coupler history output without changing HIST_N from its default.

**Comment by jedwards4b** [comment]
Is that overnight test directory on mappy - where they are apparently having problems?  If so can you try ./xmlquery HIST_N in one of those ERS tests?

**Comment by jgfouca** [comment]
To be clear, all the basics are running fine. It's only xmlquery for specific items that's the problem.

**Comment by rljacob** [comment]
No I'm trying Chrysalis.  I've seen ./xmlquery HIST_N give an error on a test dir I just created but it works fine in a test dir created by the overnight testing.  On the same machine.

**Comment by rljacob** [comment]
It must be something about that case and the SMS test.  I tried ./create_test --no-build ERP.ne256pg2_ne256pg2.F2010-SCREAMv1 and ./xmlquery HIST_N worked..

**Comment by bartgol** [comment]
> It must be something about that case and the SMS test. I tried ./create_test --no-build ERP.ne256pg2_ne256pg2.F2010-SCREAMv1 and ./xmlquery HIST_N worked..

For ERS/ERP tests, I suspect HIST_N is set to an integer, and not $STOP_N?

**Comment by jedwards4b** [comment]
Not true: https://github.com/ESMCI/cime/blob/master/CIME/data/config/config_tests.xml#L270

**Comment by rljacob** [comment]
By default, STOP_N is 5 for SMS and 11 for ERS.

**Comment by bartgol** [comment]
I can confirm that also on my laptop `./xmlquery HIST_N` works with an ERP test, but fails with an SMS one. My theory above regarding HIST_N and STOP_N being in different groups must be wrong (or incomplete).

**Comment by jasonb5** [comment]
@bartgol The issues is with `HIST_N` and `STOP_N` being in different groups. The reason `ERS` continues work is due to it having `STOP_N` defined in `env_test.xml`. Any test with `<HIST_N>$STOP_N</HIST_N>` and `<STOP_N>...</STOP_N>` defined in [config_tests.xml](https://github.com/ESMCI/cime/blob/master/CIME/data/config/config_tests.xml) will work still. But anything relying on `$STOP_N` being populated from somewhere else will fail, e.g. PEA, PEM, PET, etc.

I'm thinking the fix is probably making sure values not provided in [config_tests.xml](https://github.com/ESMCI/cime/blob/master/CIME/data/config/config_tests.xml) are given a default when the file is generated. Changing the grouping could work too, might be simpler than the first. I don't think we could remove the `subgroup` from https://github.com/ESMCI/cime/blob/67ade4a156ca53f89e118be6a272aef375d0fd08/CIME/case/case.py#L553-L557 since it'd probably re-introduce the origin issue from #4572.

**Comment by bartgol** [comment]
What if we allowed to specify indirect parameters via "scoping"? I'm thinking something like setting `HIST_N` to `$run_begin_stop_restart::STOP_N`. We'd still process $-containing strings as params to be resolved, but we'd add an extra step:

```python
        for m in reference_re.finditer(item_data):
            var = m.groups()[0]
            logger.debug("find: {}".format(var))
            # The overridden versions of this method do not simply return None
            # so the pylint should not be flagging this
            # pylint: disable=assignment-from-none
            tokens = var.split('::')
			if len(tokens)==2:
               ref = self.get_value(tokens[0], subgroup=tokens[1])
			else:
               ref = self.get_value(var, subgroup=subgroup)
```
I used the `::` token since I'm thinking about C++ scopes, but any other "safe" string would do (double underscore, @, ...)
